### ISSUE TITLE
15.2-LLL: Mitigating Catastrophic Forgetting

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#81]
三個 Lifelong Learning Catastrophic Forgetting 的可能解法
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/568a1da9-816f-44f9-8a46-a8592a558ac4)

### Selective Synaptic Plasticity
Synaptic = 突觸 = 神經跟神經之間的連結
只讓類神經網路中某一些神經元間的連結具有可塑性 
有一些連結必須被固化 它必須不能夠再移動 = Regularization-based

Selective Synaptic Plasticity 是發展得最完整的一個研究
作業裡面主要的問題 也都集中在跟 Regularization-based 有關的方法上面

假設有任務一跟任務二
假設我們的模型只有兩個參數 θ1 跟 θ2
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/27908485-6246-4601-97b0-2c2e11223c43)
顏色越偏藍色 代表 Loss 越小
顏色越偏白色 代表 Loss 越大
左右兩張圖分別就是任務一跟任務二的 Loss Function 也就是他們的 Error Surface

先讓模型訓練任務一 
你有一個隨機初始化的參數 θ0
用 Gradient Descent 的方法 去調整 θ0 的參數得到 θb

接下來我們得繼續解任務二 
把 θb同樣的參數拷貝過來 拷貝到任務二的這個 Error Surface 上面

那在任務二上 我們有另外一個不一樣的 Error Surface
根據這個任務二的 Error Surface 去再 Update 參數
可能會把 θb 往右上角移得到 θ* 
θ* 是訓練完任務一 接下來又訓練完任務二依序訓練兩個任務以後所得到的參數
現在用 θ* 來代表依序訓練完兩個任務以後所得到的參數 
θ*它在任務二上 Loss 比較低的位置 所以它在任務二上會得到好的表現
但如果你回頭再把 θ*拿回到任務一上去做使用
你會發現你並沒有辦法得到好的結果 它在任務一上不見得會有低的 Loss
這個就是 Forget 這件事情產生的原因

也許在這個藍色橢圓形範圍內 Loss 都算是夠低的
如果我們 θb 移動的方向 不要往右上移 而是往右邊移也許就不會有 Forget 的情形
所以這邊的基本的想法是說
每一個參數對我們過去學過的任務 它的重要性是不一樣的
希望在學新的參數的時候 舊的重要的參數 它的值盡量不要變
新的任務只去改那些 對過去的任務不重要的參數就好

假設 θb 是在前一個任務所學出來的參數
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/38390c1a-86cf-4041-9e1c-287d6124841f)

所以 θb 在前一個任務上是好的
那我們讓 θb 在第二個任務上繼續做學習

在 Selective Synaptic Plasticity  我們會給每一個參數一個保鏢(守衛) = bi
對每一個參數 θi 我們都有一個守衛 bi 
每一個參數就是Neural 裡面的每一個 Weight 跟它的 Bias 
如果你的 network 有 100 萬個參數的話那就有 100 萬個 bi 的值 

這個守衛代表說這個參數 對過去的任務而言 到底重不重要
我們有一個新的 Loss FunctionL'

如果 bi 的值很大 就代表說我們希望 θi 跟 θbi 非常靠近
如果 bi 的值很小 代表我們覺得 θi 沒有跟 θbi 很靠近 也無所謂

所有的 bi 我們都設為 0 那意味著一般的 Training->有 Catastrophic Forgetting 的問題
我們完全沒有要求新的 Learn 出來的參數跟過去學出來的參數有什麼樣的關係
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/e6a3ea59-8ea4-4323-ad17-50d49ec07454)

那我們就把 bi 給它設一個非常大的值你會進入另外一個極端 這個極端叫做 Intransigence = 不肯妥協 不肯讓步

### bi的設定
在文獻上  bi 都是人為設定的
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/fb14458c-988d-4e45-842b-da03ba6b8f0a)

你恐怕不能讓 bi 用 Learn 的
如果你讓 bi 自己學 對它來說它要讓 Loss 越小越好->直接 bi 等於 0

在訓練完一個模型之後我們得到 θb
我們把 θb 在 θ1 這個方向做一下移動
如果發現說 在 θ1 這個方向上做移動 好像對 Loss 沒有什麼影響
那我們就知道說 θ1 沒有很重要 θ1 對任務一沒有很重要

那既然 θ1 對任務一沒有很重要
在新的任務上θ1 就可以任意改變
所以我們就可以給 θ1 比較小的 bi 的值

反過來說 如果我們觀察 θ2
當我們改變 θ2 的值的時候 對 Loss 的影響是大的
代表說 θ2 是一個很重要的參數
θ2 對 Task 1 是重要的參數
所以你要把 θ2 的 b 設得大一點
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/03bd895b-cd37-4ec9-b31a-88f59e7ee991)

文獻上真正的實驗結果
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/c870ecaa-ddc7-479a-b30e-f5249647f798)

橫軸代表依序訓練的過程
在第一個虛線左邊指的是訓練任務 A 
我們有 A B C 三個任務 我們先訓練任務 A 然後呢 再訓練任務 B 再訓練任務 C

有三個縱軸
第一個縱軸代表任務 A 的正確率
第二個縱軸是任務 B 的正確率
第三個縱軸是任務 C 的正確率

藍色這一條線就是一般的 Training 也就是 bi 永遠都設為 0
L2 這個實驗是 bi 不管哪個參數通通設 1 
如果 bi 不管哪個參數通通設 1確實有達到防止 Catastrophic Forgetting 的效果

我們來看一下 任務 B 跟任務 C 學習的狀況 綠色這一條線
當 bi 永遠等於 1 的時候 我們在學任務 B 的時候任務 B 的正確率 卻沒有升得足夠高 -> Intransigence

任務 C 更慘-> Intransigence
那這個對你的模型來說限制太大了 會導致它新的任務學不起來 ->Intransigence

紅色這條線給不同的參數 不同的 bi

各種不同算 bi 的方法
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1a60ea25-e820-4727-9555-dc5a611e87b2)

___________

接下來呢另外兩個做法非常快地帶過去 

### Additional Neural Resource Allocation
改變每一個任務裡面的 Neural 的 Resource
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/9f4e548c-35f4-498d-a26f-e113af7763ee)
